### PR TITLE
s46307266 Improved Unet Pull Request

### PR DESCRIPTION
**Pull request by kd9911**
### Improved Unet
Hi Shakes,

This is my improved UNET based on the paper provided:

F. Isensee, P. Kickingereder, W. Wick, M. Bendszus, and K. H. Maier-Hein, “Brain Tumor Segmentation
and Radiomics Survival Prediction: Contribution to the BRATS 2017 Challenge,” Feb. 2018. [Online].
Available: https://arxiv.org/abs/1802.10508v1

This version of UNET is very similar to the one from demo 2 with as few *small* changes; Notably being, the abstraction of the model utilizing a more modular approach for the blocks of convolution which actually helped with implementation in the code. The other main difference was this model uses an elementwise sum operation on all of the localisation outputs after they have been convolved into 1 channel. Other changes to note are the inclusion of dropout layers and in my interpretation I have used Conv Transpose to upscale instead.

The challenge was to segment moles from the ISIC melanoma dataset and to achieve a dice score of 0.8 or above on the ground truth and prediction masks (for each class)

DSC scores for the skin class reached~0.93 but obviously this isn't as desired as the mole DSC scores. unfortunately I wasn't able to get 0.8 DSC score for the moles but came somewhat close at ~0.76. I found the greatest improvement was by changing the model into an implementation that uses colour (using grayscale the score were sitting around ~0.71 to 0.72).  My guess is this is because colour is an feature important when spotting moles.

Also implementing DSC as an actual loss function made the training more robust and consistent; As well as reducing the learning rate to avoid the loss stagnating. In addition, I couldn't get the model to work with Leaky ReLU or even ReLU so had to settle with the sigmoid activation function.

Besides all that, The code works well and operates correctly, it is also fairly quick ~20 min. One thing I did notice was making the epochs go above ~100 actually caused the dice scores to go down rather than up, possibly over fitting? 

Possible improvements that can be made:

- use a learning rate schedule
- get implementation to work with ReLU or Leaky ReLU (preferred) 
- Batch normalisation or Instance normalisation
- Regularization methods to reduce overfitting
- Other methods or procedures to achieve > 0.8 DSC score for moles 

Thanks,

Korbin Daly

### CODE REVIEW

**Comment by wooboyeong** [comment]
Code
* Code OK for EASY
* Reached 0.76 DSC - might be helpful to plot loss curve to observe training behaviour

General Comments
* Commit messages OK
* Documenation OK
* Pull request OK

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
No Header blocks -1
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage 
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch)
No feedback required
Request Description very good
